## load_digits : 손글씨를 분류해 봅시다

In [1]:
import pandas as pd
import sklearn

print(sklearn.__version__)

1.0.2


In [2]:
from sklearn.datasets import load_digits
from sklearn.metrics import classification_report

In [3]:
#문제지
hand_wr = load_digits()
display(hand_wr.data)
display(len(hand_wr.data))

array([[ 0.,  0.,  5., ...,  0.,  0.,  0.],
       [ 0.,  0.,  0., ..., 10.,  0.,  0.],
       [ 0.,  0.,  0., ..., 16.,  9.,  0.],
       ...,
       [ 0.,  0.,  1., ...,  6.,  0.,  0.],
       [ 0.,  0.,  2., ..., 12.,  0.,  0.],
       [ 0.,  0., 10., ..., 12.,  1.,  0.]])

1797

In [4]:
len(hand_wr.data[0])

64

In [5]:
hand_wr.data[0]

array([ 0.,  0.,  5., 13.,  9.,  1.,  0.,  0.,  0.,  0., 13., 15., 10.,
       15.,  5.,  0.,  0.,  3., 15.,  2.,  0., 11.,  8.,  0.,  0.,  4.,
       12.,  0.,  0.,  8.,  8.,  0.,  0.,  5.,  8.,  0.,  0.,  9.,  8.,
        0.,  0.,  4., 11.,  0.,  1., 12.,  7.,  0.,  0.,  2., 14.,  5.,
       10., 12.,  0.,  0.,  0.,  0.,  6., 13., 10.,  0.,  0.,  0.])

In [6]:
hand_wr.target

array([0, 1, 2, ..., 8, 9, 8])

In [7]:
from collections import Counter

Counter(hand_wr.target)

Counter({0: 178,
         1: 182,
         2: 177,
         3: 183,
         4: 181,
         5: 182,
         6: 181,
         7: 179,
         8: 174,
         9: 180})

In [8]:
display(hand_wr.target_names)
display(len(hand_wr.target_names))

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

10

In [9]:
display(hand_wr.feature_names[:5])
display(len(hand_wr.feature_names))

['pixel_0_0', 'pixel_0_1', 'pixel_0_2', 'pixel_0_3', 'pixel_0_4']

64

In [10]:
hdf = pd.DataFrame(hand_wr.data)
hdf

0    1     2     3     4     5    6    7    8    9   ...   54   55  \
0     0.0  0.0   5.0  13.0   9.0   1.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
1     0.0  0.0   0.0  12.0  13.0   5.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
2     0.0  0.0   0.0   4.0  15.0  12.0  0.0  0.0  0.0  0.0  ...  5.0  0.0   
3     0.0  0.0   7.0  15.0  13.0   1.0  0.0  0.0  0.0  8.0  ...  9.0  0.0   
4     0.0  0.0   0.0   1.0  11.0   0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
...   ...  ...   ...   ...   ...   ...  ...  ...  ...  ...  ...  ...  ...   
1792  0.0  0.0   4.0  10.0  13.0   6.0  0.0  0.0  0.0  1.0  ...  4.0  0.0   
1793  0.0  0.0   6.0  16.0  13.0  11.0  1.0  0.0  0.0  0.0  ...  1.0  0.0   
1794  0.0  0.0   1.0  11.0  15.0   1.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
1795  0.0  0.0   2.0  10.0   7.0   0.0  0.0  0.0  0.0  0.0  ...  2.0  0.0   
1796  0.0  0.0  10.0  14.0   8.0   1.0  0.0  0.0  0.0  2.0  ...  8.0  0.0   

       56   57   58    59    60    61   62   63  
0     0.0  0.0  6.0  13.0  10.0   0.0  0.0  0.0  
1     0.0  0.0  0.0  11.0  16.0  10.0  0.0  0.0  
2     0.0  0.0  0.0   3.0  11.0  16.0  9.0  0.0  
3     0.0  0.0  7.0  13.0  13.0   9.0  0.0  0.0  
4     0.0  0.0  0.0   2.0  16.0   4.0  0.0  0.0  
...   ...  ...  ...   ...   ...   ...  ...  ...  
1792  0.0  0.0  2.0  14.0  15.0   9.0  0.0  0.0  
1793  0.0  0.0  6.0  16.0  14.0   6.0  0.0  0.0  
1794  0.0  0.0  2.0   9.0  13.0   6.0  0.0  0.0  
1795  0.0  0.0  5.0  12.0  16.0  12.0  0.0  0.0  
1796  0.0  1.0  8.0  12.0  14.0  12.0  1.0  0.0  

[1797 rows x 64 columns]

#### (4) train, test 데이터 분리

In [11]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix

X_train, X_test, y_train, y_test = train_test_split(hand_wr.data,hand_wr.target,test_size=0.2,random_state=2022 )



#### (5) 다양한 모델로 학습시켜보기

In [12]:
#Decision Tree
from sklearn.metrics import accuracy_score
from sklearn.tree import DecisionTreeClassifier
decision_tree = DecisionTreeClassifier(random_state=2022)

decision_tree.fit(X_train,y_train)
y_pred = decision_tree.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
display(accuracy)
print(classification_report(y_test, y_pred))

confusion_matrix(y_test, y_pred)




0.8305555555555556

              precision    recall  f1-score   support

           0       1.00      0.89      0.94        37
           1       0.79      0.71      0.75        42
           2       0.78      0.88      0.83        33
           3       0.71      0.79      0.75        38
           4       0.73      0.90      0.80        39
           5       0.97      0.86      0.92        44
           6       0.86      0.89      0.87        35
           7       0.88      0.79      0.84        29
           8       0.83      0.77      0.80        31
           9       0.81      0.81      0.81        32

    accuracy                           0.83       360
   macro avg       0.84      0.83      0.83       360
weighted avg       0.84      0.83      0.83       360



array([[33,  1,  0,  0,  0,  0,  3,  0,  0,  0],
       [ 0, 30,  2,  4,  2,  1,  1,  0,  2,  0],
       [ 0,  0, 29,  3,  0,  0,  0,  0,  1,  0],
       [ 0,  0,  2, 30,  2,  0,  0,  1,  0,  3],
       [ 0,  0,  0,  1, 35,  0,  0,  2,  0,  1],
       [ 0,  1,  0,  0,  1, 38,  1,  0,  1,  2],
       [ 0,  0,  0,  0,  4,  0, 31,  0,  0,  0],
       [ 0,  0,  0,  2,  3,  0,  0, 23,  1,  0],
       [ 0,  2,  4,  1,  0,  0,  0,  0, 24,  0],
       [ 0,  4,  0,  1,  1,  0,  0,  0,  0, 26]], dtype=int64)

In [13]:
#Random Forest
from sklearn.ensemble import RandomForestClassifier
random_forest = RandomForestClassifier(random_state=32)

random_forest.fit(X_train,y_train)
y_pred = random_forest.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        37
           1       0.95      0.98      0.96        42
           2       1.00      1.00      1.00        33
           3       1.00      0.92      0.96        38
           4       1.00      0.97      0.99        39
           5       0.98      0.93      0.95        44
           6       0.97      1.00      0.99        35
           7       0.94      1.00      0.97        29
           8       0.94      0.97      0.95        31
           9       0.94      0.97      0.95        32

    accuracy                           0.97       360
   macro avg       0.97      0.97      0.97       360
weighted avg       0.97      0.97      0.97       360



In [14]:
#SVM 
from sklearn import svm 
svm_model = svm.SVC()

svm_model.fit(X_train,y_train)
y_pred = svm_model.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        37
           1       0.98      1.00      0.99        42
           2       1.00      1.00      1.00        33
           3       1.00      0.95      0.97        38
           4       1.00      0.97      0.99        39
           5       1.00      0.93      0.96        44
           6       0.97      1.00      0.99        35
           7       0.97      1.00      0.98        29
           8       0.97      0.97      0.97        31
           9       0.91      1.00      0.96        32

    accuracy                           0.98       360
   macro avg       0.98      0.98      0.98       360
weighted avg       0.98      0.98      0.98       360



In [15]:
#SGD Classifier
from sklearn.linear_model import SGDClassifier
sgd_model = SGDClassifier()

sgd_model.fit(X_train,y_train)
y_pred = sgd_model.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        37
           1       0.93      0.88      0.90        42
           2       1.00      0.91      0.95        33
           3       0.95      0.92      0.93        38
           4       1.00      0.97      0.99        39
           5       0.98      0.93      0.95        44
           6       0.95      1.00      0.97        35
           7       0.90      0.97      0.93        29
           8       0.79      0.97      0.87        31
           9       0.97      0.91      0.94        32

    accuracy                           0.94       360
   macro avg       0.95      0.95      0.94       360
weighted avg       0.95      0.94      0.95       360



In [16]:
#Logistic Regression
from sklearn.linear_model import LogisticRegression
logistic_model = LogisticRegression()

logistic_model.fit(X_train,y_train)
y_pred = logistic_model.predict(X_test)
print(classification_report(y_test, y_pred))



              precision    recall  f1-score   support

           0       1.00      1.00      1.00        37
           1       0.97      0.93      0.95        42
           2       1.00      0.94      0.97        33
           3       0.97      0.92      0.95        38
           4       0.97      0.97      0.97        39
           5       0.95      0.91      0.93        44
           6       0.97      1.00      0.99        35
           7       0.97      0.97      0.97        29
           8       0.84      0.87      0.86        31
           9       0.84      1.00      0.91        32

    accuracy                           0.95       360
   macro avg       0.95      0.95      0.95       360
weighted avg       0.95      0.95      0.95       360



c:\Anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [17]:
display(decision_tree._estimator_type)
display(random_forest._estimator_type)
display(svm_model._estimator_type)
display(sgd_model._estimator_type)
display(logistic_model._estimator_type)

'classifier'

'classifier'

'classifier'

'classifier'

'classifier'

In [18]:
confusion_matrix(y_test, y_pred)

array([[37,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0, 39,  0,  0,  1,  0,  0,  0,  2,  0],
       [ 0,  0, 31,  1,  0,  0,  0,  0,  1,  0],
       [ 0,  0,  0, 35,  0,  1,  0,  0,  2,  0],
       [ 0,  0,  0,  0, 38,  0,  0,  0,  0,  1],
       [ 0,  0,  0,  0,  0, 40,  1,  1,  0,  2],
       [ 0,  0,  0,  0,  0,  0, 35,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0,  0, 28,  0,  1],
       [ 0,  1,  0,  0,  0,  1,  0,  0, 27,  2],
       [ 0,  0,  0,  0,  0,  0,  0,  0,  0, 32]], dtype=int64)

#### (6) 모델을 평가해 보기

In [19]:
#SVM 
from sklearn import svm 
svm_model = svm.SVC()

svm_model.fit(X_train,y_train)
y_pred = svm_model.predict(X_test)
report = classification_report(y_test, y_pred)
print(report)

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        37
           1       0.98      1.00      0.99        42
           2       1.00      1.00      1.00        33
           3       1.00      0.95      0.97        38
           4       1.00      0.97      0.99        39
           5       1.00      0.93      0.96        44
           6       0.97      1.00      0.99        35
           7       0.97      1.00      0.98        29
           8       0.97      0.97      0.97        31
           9       0.91      1.00      0.96        32

    accuracy                           0.98       360
   macro avg       0.98      0.98      0.98       360
weighted avg       0.98      0.98      0.98       360



In [20]:
confusion_matrix(y_test, y_pred)

array([[37,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0, 42,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0, 33,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0, 36,  0,  0,  0,  1,  1,  0],
       [ 0,  0,  0,  0, 38,  0,  0,  0,  0,  1],
       [ 0,  0,  0,  0,  0, 41,  1,  0,  0,  2],
       [ 0,  0,  0,  0,  0,  0, 35,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0,  0, 29,  0,  0],
       [ 0,  1,  0,  0,  0,  0,  0,  0, 30,  0],
       [ 0,  0,  0,  0,  0,  0,  0,  0,  0, 32]], dtype=int64)

In [21]:
len(y_test)

360

In [22]:
y_pred

array([6, 9, 8, 0, 8, 2, 6, 3, 3, 5, 4, 5, 8, 4, 8, 1, 6, 1, 3, 2, 6, 8,
       0, 6, 7, 9, 9, 6, 4, 0, 9, 4, 2, 1, 4, 4, 8, 9, 2, 9, 1, 2, 9, 8,
       9, 0, 0, 1, 9, 8, 1, 0, 5, 6, 2, 9, 8, 6, 0, 4, 7, 2, 4, 4, 2, 4,
       5, 1, 3, 0, 4, 4, 3, 3, 6, 6, 2, 5, 2, 0, 9, 4, 0, 7, 1, 8, 0, 0,
       2, 7, 1, 4, 1, 6, 7, 7, 6, 2, 1, 7, 9, 9, 3, 7, 5, 5, 0, 0, 6, 0,
       6, 7, 9, 4, 7, 1, 0, 6, 0, 7, 0, 3, 6, 9, 6, 9, 1, 7, 2, 1, 9, 9,
       5, 5, 9, 8, 4, 7, 5, 0, 6, 8, 5, 0, 5, 9, 5, 6, 5, 2, 7, 4, 7, 2,
       1, 0, 3, 3, 3, 5, 2, 1, 5, 7, 8, 2, 2, 1, 3, 8, 6, 0, 3, 3, 9, 5,
       9, 1, 4, 4, 2, 0, 7, 3, 7, 1, 5, 7, 8, 3, 2, 8, 5, 8, 5, 1, 4, 1,
       1, 9, 4, 9, 5, 4, 5, 6, 6, 7, 5, 6, 8, 7, 1, 6, 4, 5, 2, 0, 6, 3,
       6, 0, 0, 7, 7, 4, 1, 8, 4, 3, 3, 9, 5, 8, 1, 1, 6, 6, 2, 2, 1, 7,
       3, 5, 3, 2, 3, 1, 5, 9, 8, 9, 0, 4, 1, 4, 0, 7, 6, 3, 5, 9, 0, 2,
       0, 3, 5, 3, 5, 3, 2, 3, 9, 3, 5, 1, 9, 1, 0, 1, 1, 2, 8, 9, 4, 4,
       5, 6, 0, 3, 6, 8, 9, 7, 2, 5, 8, 0, 2, 7, 5,

In [23]:
import numpy as np

new_df = np.array([0]*360)
new_df = pd.DataFrame(y_pred)

In [24]:
new_df['y_test'] = y_test
new_df

0  y_test
0    6       6
1    9       9
2    8       8
3    0       0
4    8       8
..  ..     ...
355  4       4
356  4       4
357  1       1
358  3       3
359  2       2

[360 rows x 2 columns]

In [25]:
new_df = new_df.rename(columns={0:'y_pred'})

new_df['minus'] = new_df['y_pred']-new_df['y_test']
new_df

y_pred  y_test  minus
0         6       6      0
1         9       9      0
2         8       8      0
3         0       0      0
4         8       8      0
..      ...     ...    ...
355       4       4      0
356       4       4      0
357       1       1      0
358       3       3      0
359       2       2      0

[360 rows x 3 columns]

In [26]:
new_df[new_df['minus'] != 0]

y_pred  y_test  minus
26        9       5      4
37        9       4      5
48        9       5      4
57        6       5      1
89        7       3      4
167       1       8     -7
233       8       3      5

숫자가 5인데, 6이나 9로 예측한 경우가 많다.. ^^; 

In [27]:
#SVM 
from sklearn import svm 
svm_model = svm.SVC()

svm_model.fit(X_train,y_train)
y_pred = svm_model.predict(X_test)
report = classification_report(y_test, y_pred)
print(report)

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        37
           1       0.98      1.00      0.99        42
           2       1.00      1.00      1.00        33
           3       1.00      0.95      0.97        38
           4       1.00      0.97      0.99        39
           5       1.00      0.93      0.96        44
           6       0.97      1.00      0.99        35
           7       0.97      1.00      0.98        29
           8       0.97      0.97      0.97        31
           9       0.91      1.00      0.96        32

    accuracy                           0.98       360
   macro avg       0.98      0.98      0.98       360
weighted avg       0.98      0.98      0.98       360



프로젝트1에서 글씨를 분류하기에 적합한 모델은 SVM이라고 할 수 있다. 
f1-score 점수가 가장 높다. 

틀린비율과 맞는 비율 모두 골고루 섞인 f1-score에서 weighted avg까지 고려한.. 값이 0.98

아무튼 최고당,.